File: common/src/main/java/net/shiroha233/roadweaver/features/path/decoration/system/RoadBlockPlacer.java
package net.shiroha233.roadweaver.features.path.decoration.system;

import net.minecraft.core.BlockPos;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.WorldGenLevel;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.BlockState;
import net.shiroha233.roadweaver.config.ModConfig;

import java.util.List;

/**
 * Road block placement with controlled material variation.
 *
 * The first palette entry is treated as the local biome's primary material and
 * is deliberately favoured. Secondary materials add visual variation without
 * turning the road into noisy random block spam.
 */
public final class RoadBlockPlacer {
    private RoadBlockPlacer() {}

    public static void placeRoadBlock(WorldGenLevel world,
            BlockState blockBelow,
            BlockPos surfacePos,
            List<BlockState> materials,
            RandomSource random,
            ModConfig cfg) {
        if (!PlacementRules.placeAllowedCheck(blockBelow.getBlock()) || materials == null || materials.isEmpty())
            return;

        BlockState chosen = chooseMaterial(materials, random);
        BlockPos below1 = surfacePos.below();
        world.setBlock(below1, chosen, 3);

        AboveColumnClearer.clearAboveColumn(world, surfacePos, cfg);

        BlockPos belowPos1 = surfacePos.below(2);
        BlockState belowState1 = world.getBlockState(belowPos1);
        if (belowState1.is(Blocks.GRASS_BLOCK)) {
            world.setBlock(belowPos1, Blocks.DIRT.defaultBlockState(), 3);
        }
    }

    private static BlockState chooseMaterial(List<BlockState> materials, RandomSource random) {
        if (materials.size() == 1) {
            return materials.get(0);
        }

        // Keep the dominant surface material visually coherent. The remaining
        // 30% is spread across secondary materials for natural variation.
        if (random.nextInt(100) < 70) {
            return materials.get(0);
        }
        return materials.get(1 + random.nextInt(materials.size() - 1));
    }
}

File: common/src/main/java/net/shiroha233/roadweaver/features/path/decoration/system/RoadBlockPlacer.kt
package net.shiroha233.roadweaver.features.path.decoration.system

import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.util.RandomSource
import net.minecraft.world.level.WorldGenLevel
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.state.BlockState
import net.shiroha233.roadweaver.config.ModConfig
import kotlin.math.max
import kotlin.math.min

object RoadBlockPlacer {
    @JvmStatic
    fun placeRoadBlock(
        world: WorldGenLevel,
        blockBelow: BlockState,
        surfacePos: BlockPos,
        materials: List<BlockState>,
        random: RandomSource,
        cfg: ModConfig?
    ) {
        if (!PlacementRules.placeAllowedCheck(blockBelow.block)) return
        val chosen = materials[random.nextInt(materials.size)]

        val MAX_CAUSEWAY_DEPTH = max(0, min(12, cfg?.causewayMaxDepth() ?: 1))
        val below1 = surfacePos.below()
        val below2 = surfacePos.below(2)
        val sturdy1 = world.getBlockState(below1).isFaceSturdy(world, below1, Direction.UP)
        val sturdy2 = world.getBlockState(below2).isFaceSturdy(world, below2, Direction.UP)

        if (!sturdy1 && !sturdy2) {
            var cursor = below2
            var depth = 0
            var base: BlockPos? = null
            while (cursor.y > world.minBuildHeight && depth < MAX_CAUSEWAY_DEPTH) {
                if (world.getBlockState(cursor).isFaceSturdy(world, cursor, Direction.UP)) {
                    base = cursor
                    break
                }
                cursor = cursor.below()
                depth++
            }

            var fillStart = base?.above()
                ?: below1.below(min(MAX_CAUSEWAY_DEPTH - 1, max(0, below1.y - world.minBuildHeight)))
            if (fillStart.y < world.minBuildHeight) {
                fillStart = BlockPos(fillStart.x, world.minBuildHeight, fillStart.z)
            }

            var pos = fillStart
            while (pos.y <= below1.y) {
                world.setBlock(pos, chosen, 3)
                pos = pos.above()
            }
        } else {
            world.setBlock(below1, chosen, 3)
        }

        AboveColumnClearer.clearAboveColumn(world, surfacePos, cfg)

        val belowPos1 = surfacePos.below(2)
        val belowState1 = world.getBlockState(belowPos1)
        if (belowState1.`is`(Blocks.GRASS_BLOCK)) {
            world.setBlock(belowPos1, Blocks.DIRT.defaultBlockState(), 3)
        }
    }
}
